store/tikv/2pc.go

```diff
@@ -422,6 +414,24 @@ func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) er
 	return nil
 }
 
+func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) error {
+	// If we fail to receive response for the request that commits primary key, it will be undetermined whether this
+	// transaction has been successfully committed.
+	// Under this circumstance,  we can not declare the commit is complete (may lead to data lost), nor can we throw
+	// an error (may lead to the duplicated key error when upper level restarts the transaction). Currently the best
+	// solution is to populate this error and let upper layer drop the connection to the corresponding mysql client.
+	isPrimary := false
+	if bytes.Compare(batch.keys[0], c.primary()) == 0 {
+		isPrimary = true
+	}
+
+	if err := c.doCommitSingleBatch(bo, batch); err != nil && isPrimary {
+		// change the Cause of the error to be returned
+		return errors.Wrap(err, terror.ErrResultUndetermined)
+	}
+	return nil
```

Comment:
@tiancaiamao The context for executing the tikv tasks is not related to the context of client connection. So the error handling here is not effected by manual cancellation of either query or connection.